### PR TITLE
Add menu2 recycler view adapter

### PR DESCRIPTION
**Pull request by NotWoods**
`MenuCandidateListAdapter` converts MenuCandidate items into a list of menu items. Probably easiest to start with that class!

Split from #5496

---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/MenuIconAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Different holder classes are used to swap out the child views.
+ */
+internal class MenuIconAdapter(
+    private val parent: ConstraintLayout,
+    private val inflater: LayoutInflater,
+    private val side: Side,
+    private val dismiss: () -> Unit
+) {
+
+    private var viewHolder: MenuIconViewHolder<*>? = null
```
Comment (jonalmeida) [inline components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/MenuIconAdapter.kt:27]:
```suggestion
    private var viewHolder: MenuIconViewHolder<out MenuIcon>? = null
```

Same comment here I think.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/MenuIconViewHolder.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * View holder with a [bind] method that passes the previously bound value.
+ */
+internal abstract class MenuIconViewHolder<T : MenuIcon>(
+    protected val parent: ConstraintLayout,
+    protected val inflater: LayoutInflater
+) {
+
+    @Suppress("Unchecked_Cast")
+    fun bindAndCast(newIcon: MenuIcon, oldIcon: MenuIcon?) {
+        bind(newIcon as T, oldIcon as T?)
```
Comment (jonalmeida) [inline components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/MenuIconViewHolder.kt:28]: That's one of the unfortunate side-effects of using sealed classes. I don't have any simple solutions here besides trying to make `MenuIcon` an interface, but that might pose larger changes in how you match the view holders.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/View.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.menu2.ext
```
Comment (jonalmeida) [inline components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/View.kt:5]: Test for all of these please. :)

File: components/browser/menu2/src/main/res/values/colors.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<resources>
+    <!-- Empty by default, allows others to theme as they see fit -->
```
Comment (jonalmeida) [inline components/browser/menu2/src/main/res/values/colors.xml:6]: We have some other components that use styles to allow the app to set a colour here. The [BrowserTabsTray](https://github.com/mozilla-mobile/android-components/blob/40d97f22a8639849297c8bf6099ed23871a3cdbc/components/browser/tabstray/src/main/java/mozilla/components/browser/tabstray/BrowserTabsTray.kt#L43) is an example off the top of my mind you could check out.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/MenuCandidateListAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            is DividerMenuCandidateViewHolder -> holder.bind(item as DividerMenuCandidate)
+        }
+    }
+}
+
+private object MenuCandidateDiffer : DiffUtil.ItemCallback<MenuCandidate>() {
+    override fun areItemsTheSame(oldItem: MenuCandidate, newItem: MenuCandidate) =
+        oldItem::class == newItem::class
+
+    @Suppress("DiffUtilEquals")
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/MenuCandidateListAdapter.kt:68]: Using sealed classes throws this error sometimes. I'll try removing and see if anything changes.

**Comment by NotWoods** [comment]
The original PR is #5496. I hooked up Sample Browser in that PR to use the new system.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/RowMenuCandidateViewHolder.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    inflater: LayoutInflater,
+    private val dismiss: () -> Unit
+) : MenuCandidateViewHolder<RowMenuCandidate>(itemView, inflater) {
+
+    private val layout = itemView as LinearLayout
+    private var buttonViewHolders = emptyList<SmallMenuCandidateViewHolder>()
+
+    override fun bind(newCandidate: RowMenuCandidate, oldCandidate: RowMenuCandidate?) {
+        super.bind(newCandidate, oldCandidate)
+        if (newCandidate.items.size != oldCandidate?.items?.size) {
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/RowMenuCandidateViewHolder.kt:25]: The code after the if statement does a deeper diff on the children. This is a quick check to see if the number of children need to change too.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        updateConstraints {
+            connect(R.id.icon, TOP, PARENT_ID, TOP)
+            connect(R.id.icon, BOTTOM, PARENT_ID, BOTTOM)
+            val margin = parent.resources
+                .getDimensionPixelSize(R.dimen.mozac_browser_menu2_icon_padding_start)
+            when (side) {
+                Side.START -> {
+                    connect(imageView.id, START, PARENT_ID, START)
+                    connect(imageView.id, END, R.id.label, START, margin)
+                    connect(R.id.label, START, imageView.id, END)
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt:45]: This is a feature called a [chain](https://constraintlayout.com/basics/create_chains.html). I don't think there are any negative issues with connecting from both sides, behaviour is more predictable (margins are obeyed on both views) and more functionality is enabled.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    }
+}
+
+internal class DrawableMenuIconViewHolder(
+    parent: ConstraintLayout,
+    inflater: LayoutInflater,
+    side: Side
+) : MenuIconWithDrawableViewHolder<DrawableMenuIcon>(parent, inflater) {
+
+    override val imageView: ImageView = inflate(layoutResource).findViewById(R.id.icon)
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt:68]: `ImageView` is a base class, but I could change the type here to the app compat version.

File: components/browser/menu2/src/main/res/values/colors.xml (hunk 4 above)
Comment (NotWoods) [reply components/browser/menu2/src/main/res/values/colors.xml:6]: I'm going to remove this file anyways, it only matters for the other PR.

File: components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/DrawableMenuIconViewHolder.kt]: Neat, I've never done this programmatically before. 🙂

**Comment by jonalmeida** [comment]
> The original PR is #5496. I hooked up Sample Browser in that PR to use the new system.

I see the Sample Toolbar uses it. I can check that out, although I think our fully-fledged browsers make more dynamic use of the toolbar, so it would be worth testing that out as well if you haven't.

**Comment by jonalmeida** [comment]
Don't forget to update the changelog when you land this. :)

**Comment by NotWoods** [comment]
Since the feature still isn't usable on its own, I don't want to add it in the changelog yet. I'll fix the issue and merge

**Comment by NotWoods** [comment]
bors r=jonalmeida

**Comment by NotWoods** [comment]
bors retry

**Pull request merged**